xwp/stream
# Use scheduled action for resetting large record and meta tables

Fixes #1503 

# Development notes

## Changes to default database reset functionality:

 - if the site is a single site installation or multisite install with the plugin network activated, the Stream and Stream meta tables are truncated and then a cleanup function is run on the meta table. (This is for the case where a record is added in between truncation the `{$wpdb->streammeta}` table and the `{$wpdb->stream}` table.)
 - if the site is a multisite which is not network activated and
   - the site has <= 1000000 records in the `{$wpdb->stream}` table: the previous deletion functionality will run
   - the site has > 1000000 records in the `{$wpdb->stream}` table:
     - an action will be scheduled async to delete 250k records at a time using the previous deletion functionality. The limit is handled by a range on the `{$wpdb->stream}.ID` column
     - the action is recursive and will run until there are no records left with an ID lower than the initial ID. This is to avoid the situation where records keep getting added as the actions run so they never finish.

The "Reset Stream Database" link in Stream --> Settings --> Advanced will not show while the records are being deleted. The warning given will depend on whether or not it's going to be an asynchronous process.

In my local environment, each pass of deleting 250k records takes 10-15 seconds. 

## Added filters

- `wp_stream_is_large_records_table`: filter the number of records to be processed when deciding what is a large records table. Default is 1M, ie anything over 1M records to be processed is considered a large records table.
- `wp_stream_batch_size`: how many records in the `{$wpdb->stream}` table will be used when doing the deletion.

## Incidental additions:
 - `$this->plugin->get_site_type()` returns what sort of site it is:
  - `Admin::WP_STREAM_MULTI_NETWORK` is a multisite with the plugin network activated
  - `Admin::::WP_STREAM_MULTI_NOT_NETWORK` is a multisite where the plugin is not network enabled
  - `Admin::WP_STREAM_SINGLE_SITE` is a single site
- [Action Scheduler library ](https://github.com/woocommerce/action-scheduler) to use instead of WP cron functions

## Testing instructions:

You can use the `npm run large-records-generate` to generate a bunch of records. You can see how many you've generated by running `npm run large-records-show`


# Checklist

- [x] Project documentation has been updated to reflect the changes in this pull request, if applicable.
- [x] I have tested the changes in the local development environment (see `contributing.md`).
- [x] I have added phpunit tests.


## Release Changelog

- Fix: Use async process for resetting the Stream database on large multisites where the plugin is not network enabled
- Fix: Truncate tables when resetting the Stream database on single sites and multisites where the plugin is network enabled
